controllers/phases.go

```diff
@@ -0,0 +1,477 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	versionutil "k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/pointer"
+	operatorv1 "sigs.k8s.io/cluster-api-operator/api/v1alpha1"
+	"sigs.k8s.io/cluster-api-operator/controllers/genericprovider"
+	"sigs.k8s.io/cluster-api-operator/util"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	configclient "sigs.k8s.io/cluster-api/cmd/clusterctl/client/config"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/repository"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/yamlprocessor"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const metadataFile = "metadata.yaml"
+
+// phaseReconciler holds all required information for interacting with clusterctl code and
+// helps to iterate through provider reconciliation phases.
+type phaseReconciler struct {
+	provider     genericprovider.GenericProvider
+	providerList genericprovider.GenericProviderList
+
+	ctrlClient           client.Client
+	ctrlConfig           *rest.Config
+	certManagerInstaller SingletonInstaller
+	repo                 repository.Repository
+	contract             string
+	options              repository.ComponentsOptions
+	providerConfig       configclient.Provider
+	configClient         configclient.Client
+	components           repository.Components
+	clusterctlProvider   *clusterctlv1.Provider
+}
+
+// reconcilePhaseFn is a function that represent a phase of the reconciliation.
+type reconcilePhaseFn func(context.Context) (reconcile.Result, error)
+
+// PhaseError custom error type for phases.
+type PhaseError struct {
+	Reason   string
+	Type     clusterv1.ConditionType
+	Severity clusterv1.ConditionSeverity
+	Err      error
+}
+
+func (p *PhaseError) Error() string {
+	return p.Err.Error()
+}
+
+func ifErrorWrapPhaseError(err error, reason string, ctype clusterv1.ConditionType) error {
```

Comment:
The name of this is odd to me, why `ifError`, I don't think that adds anything, especially as we only use it when we are sure the `err != nil`, so perhaps we should strip that prefix